### PR TITLE
Group references by FileId

### LINKED ISSUE #4901: Incorrect `textDocument/rename` response when multiple edits on same document.
I just noticed that when issuing a `textDocument/rename` request on a file that has N occurrences of the symbol to be renamed, rust-analyzer replies with N `TextDocumentEdit`s with a single `TextEdit` in it, whereas other servers reply with a single `TextDocumentEdit` with N `TextEdit`s.

> A TextDocumentEdit describes all changes on a version Si and after they are applied move the document to version Si+1. 

This, according to the specification (above), seems incorrect to me, as applying a `TextDocumentEdit` effectively changes the version of the document, and any subsequent applies to the older version are potentially not valid anymore. More importantly it states that a `TextDocumentEdit` describes **all** changes on a version.

This is affecting LanguageClient-neovim right now, as filed in https://github.com/autozimu/LanguageClient-neovim/issues/1055, when there are two occurrences of that symbol in the same line. If the leftmost edit is applied first, then the range for the rightmost edit is invalid, as the document has changed. When all the edits for a document are listed in a single `TextDocumentEdit`, `LanguageClient-neovim` sorts those edits by character and applies them from right to left, but given the current response this is not possible, as there is a single edit for each text document (except all text documents are really the same).

For what it's worth, we could potentially fix this in LanguageClient-neovim by grouping the `TextDocumentEdit`s by text document URI, but I feel like this is just a misinterpretation of the specification.

### PR DESCRIPTION
**Pull request by Veykril**
Fixes #4901

This doesn't address https://github.com/rust-analyzer/rust-analyzer/pull/7032/files#diff-a7e1e771e911237bb893e1b0f5e0f2c2a856b54ca06f95ef0818a922f1a8b5ebR266

### CODE REVIEW

File: crates/ide_db/src/search.rs [hunk 1]
```diff
@@ -252,23 +267,33 @@ impl<'a> FindUsages<'a> {
...
         found
     }
 
-    pub fn all(self) -> Vec<Reference> {
-        let mut res = Vec::new();
-        self.search(&mut |reference| {
-            res.push(reference);
+    /// The [`FileReferences`] returned always have unique [`FileId`]s.
+    pub fn all(self) -> Vec<FileReferences> {
+        let mut res = <Vec<FileReferences>>::new();
```
Comment (matklad) [inline crates/ide_db/src/search.rs:279]: o_O

That's a quite refined use of the turbofish! Let's stick to the more common `Vec::<FileReferences>::new();` though

File: crates/ide_db/src/search.rs [hunk 2]
```diff
@@ -252,23 +267,33 @@ impl<'a> FindUsages<'a> {
...
         });
         found
     }
 
-    pub fn all(self) -> Vec<Reference> {
-        let mut res = Vec::new();
-        self.search(&mut |reference| {
-            res.push(reference);
+    /// The [`FileReferences`] returned always have unique [`FileId`]s.
+    pub fn all(self) -> Vec<FileReferences> {
```
Comment (matklad) [inline crates/ide_db/src/search.rs:278]:
```suggestion
    pub fn all(self) -> FxHashMap<FileId, Vec<FileReference>> {
```

that way, we guarantee no duplicates. Maybe it makes sense to add

```
pub struct SearchResults {
    map: FxHashMap<FileId, Vec<FileReference>>
}
```

Comment (Veykril) [reply crates/ide_db/src/search.rs:278]: Right, not sure how using a Hashmap slipped my mind here.

**Comment by Veykril** [comment]
Alright, switched to a hashmap and wrapped that in a `UsageSearchResult` struct for some utility stuff.

**Review by matklad — COMMENTED** [review]
bors r+

File: crates/ide/src/references.rs [hunk 3]
```diff
@@ -29,7 +29,7 @@ use crate::{display::TryToNav, FilePosition, FileRange, NavigationTarget, RangeI
 #[derive(Debug, Clone)]
 pub struct ReferenceSearchResult {
     declaration: Declaration,
-    references: Vec<Reference>,
+    references: UsageSearchResult,
```
Comment (matklad) [inline crates/ide/src/references.rs:32]: Getting back to this, it is interesting. This module wants to poke the USR quite a bit. The reason for that is I think is https://github.com/rust-analyzer/rust-analyzer/issues/4290. `ReferenceSearchResult` should be a stupid POD
UI type, which is just converted to JSON and shipped to the editor. Instead, it is used by rename to drive the logic of computation. 

I think we need to change the rename to work with `ide_db/search` directly. After that, we should be able to privatize the `UsageSearchResult::references` and slim down `ide/references.rs`

**Pull request merged**